### PR TITLE
reverse_complement: rayon version

### PR DESCRIPTION
**Pull request by cristicbz**
This runs about as fast on my computer, but it's shorter and gets rid of the `Iterator` implementations (that needed a fair number `mem::replace` dances which made the code less readable IMO).

Please have a look, thanks!

### CODE REVIEW

File: src/reverse_complement.rs [hunk 1]
```diff
@@ -4,13 +4,14 @@
 // contributed by the Rust Project Developers
 // contributed by Matt Brubeck
 // contributed by TeXitoi
+// contributed by Cristi Cobzarenco (@cristicbz)
```
Comment (TeXitoi) [inline src/reverse_complement.rs]: As there is already a ticket open for this benchmark, I'll submit it by updating the current ticket. Thus, please let me be the last contributor.

File: src/reverse_complement.rs [hunk 2]
```diff
@@ -140,27 +108,14 @@ fn reverse_complement(seq: &mut [u8], tables: &Tables) {
...
-            });
-        }
-    });
+    let (left, right) = unsafe {
+        let p = seq.as_mut_ptr();
+        // This is slow if len % 2 != 0 but still faster than bytewise operations.
+        let q = p.offset((div * 2 + rem) as isize);
+        (slice::from_raw_parts_mut(p as *mut u16, div),
+         slice::from_raw_parts_mut(q as *mut u16, div))
+    };
```
Comment (TeXitoi) [inline src/reverse_complement.rs]: Then you have left, right that alias seq for the rest of the function. Maybe we can abstract this unsafe code in a safe abstraction, modeling that left and right borrow seq.
```rust
fn split_mut_middle_as_u16<'a>(&'a mut [u8]) -> (&'a mut [u16], &'a mut [u8], &'a mut [u16])
```

**Review by TeXitoi — CHANGES_REQUESTED** [review]
Great work!

**Comment by TeXitoi** [comment]
@mbrubeck can you review this PR please ?

**Comment by mbrubeck** [comment]
Looks good to me! I can also confirm that it runs as fast as the `master` branch on my computer.

File: src/reverse_complement.rs (hunk 2 above)
Comment (cristicbz) [reply src/reverse_complement.rs]: Good point, done.

File: src/reverse_complement.rs (hunk 1 above)
Comment (cristicbz) [reply src/reverse_complement.rs]: Done.

File: src/reverse_complement.rs [hunk 3]
```diff
@@ -65,61 +66,44 @@ impl Tables {
...
-            return None;
-        }
-        let n = cmp::min(self.n, len);
-        let (x, left)  = mem::replace(&mut self.left,  &mut []).split_at_mut(n);
-        let (right, y) = mem::replace(&mut self.right, &mut []).split_at_mut(len - n);
-        self.left = left;
-        self.right = right;
-        Some((x, y))
+/// Split a byte slice into two u16-halves with any remainder left in the middle.
+#[inline]
```
Comment (TeXitoi) [inline src/reverse_complement.rs]: I think we don't care of that.

**Review by TeXitoi — APPROVED** [review]

File: src/reverse_complement.rs (hunk 3 above)
Comment (cristicbz) [reply src/reverse_complement.rs]: Fair enough, doesn't make any difference.

**Comment by cristicbz** [comment]
I think this is done if you want to merge.

**Comment by TeXitoi** [comment]
https://alioth.debian.org/tracker/index.php?func=detail&aid=315633&group_id=100815&atid=413122

**Pull request merged**